generator/src/main/java/com/graphicsfuzz/generator/mutateapi/ExpressionGenerator.java

```diff
@@ -0,0 +1,632 @@
+    ;
+    final VariableDeclFact variableDeclFact = new VariableDeclFact(variablesDecl,
+        variableDeclInfo, value);
+    factManager.addVariableFact(value, variableDeclFact);
+    return new VariableIdentifierExpr(varName);
+  }
+
+  private Expr generateAdditionValue(FactManager factManager,
+                                     FunctionDefinition functionDefinition,
+                                     Stmt stmt, Value value) {
+    if (!(value instanceof NumericValue)) {
+      return null;
+    }
+
+    if (value.getType() != BasicType.INT && value.getType() != BasicType.FLOAT) {
+      return null;
+    }
+
+    final Pair<Optional<Number>, Optional<Number>> pair = getPairSum(value);
+    return new BinaryExpr(
+        generateExpr(factManager, functionDefinition, stmt,
+            new NumericValue((BasicType) value.getType(), pair.getLeft())),
+        generateExpr(factManager, functionDefinition, stmt,
+            new NumericValue((BasicType) value.getType(), pair.getRight())),
+        BinOp.ADD
+    );
+  }
+
+  private Expr generateKnownFunctionFact(FactManager factManager,
+                                         Value value,
+                                         FunctionDefinition functionDefinition,
+                                         Stmt stmt) {
+    final List<FunctionFact> availableFacts = factManager.getFunctionFacts(value);
+    if (availableFacts.isEmpty()) {
+      return null;
+    }
+    final FunctionFact functionFact = availableFacts.get(generator.nextInt(availableFacts.size()));
+    final List<Value> argValues = functionFact.getArguments();
+    final List<Expr> args = generateFunctionArgs(factManager, functionDefinition, stmt,
+        argValues);
+    return new FunctionCallExpr(functionFact.getFunctionName(), args);
+  }
+
+  private Expr generateKnownVariableFact(FactManager factManager, Value value) {
+    final List<VariableFact> availableFacts = factManager.getVariableFacts(value);
+    if (availableFacts.isEmpty()) {
+      return null;
+    }
+    final VariableFact variableFact = availableFacts.get(generator.nextInt(availableFacts.size()));
+    return new VariableIdentifierExpr(variableFact.getVariableName());
+  }
+
+  /**
+   * Utility function to generate a set of expressions which will be used as the arguments of the
+   * function call expression.
+   *
+   * @param factManager        manager class holding the value and its associated expression that
+   *                           guarantees to compute the given value.
+   * @param functionDefinition a function into which the new expression will be injected.
+   * @param stmt               statement in the body of the given function where the generated
+   *                           expression will be inserted before.
+   * @param argValues          values of function arguments from which the new values being
+   *                           generated are derived.
+   * @return a list of function argument expressions.
+   */
```

Comment:
Can you rename this to "generateArgsForFunctionCall" to make it clearer that these are the expression used to invoke the function (rather than the function's formal parameters)?

Reply:
I think you could give "stmt" a better name, e.g. "stmtToInsertBefore".